### PR TITLE
=str Make Source.completionStage can cancel the origin completionStage.

### PR DESCRIPTION
**Pull request by He-Pin**
refs: https://github.com/akka/akka/issues/31622
@johanandren Do you think this is acceptable? Tests is still missing for now, open up for a discussion.

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/impl/fusing/CompletionStageSource.scala [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        case scala.util.Success(null) => completeStage()
+        case scala.util.Success(v)    => emit(out, v, () => completeStage())
+        case scala.util.Failure(t)    => failStage(t)
+      }
+
+      override def onDownstreamFinish(cause: Throwable): Unit = {
+        super.onDownstreamFinish(cause)
+        stage match {
+          case future: Future[T @unchecked] =>
+            future.cancel(true)
```
Comment (He-Pin) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/CompletionStageSource.scala:68]: canceling happens here

**Comment by johanandren** [comment]
I don't think the value of it is big enough to warrant maintaining a separate operator implementation, and also, if we do it here we would need to do it for all operators accepting a returned `CompletionStage` for consistency, and that is definitely not worth it.

**Comment by He-Pin** [comment]
I think the  `Source.fromCompletionStage` is too common for just a single value, such as `Mono.fromCompletionStage`, maybe this one is acceptable?

**Comment by He-Pin** [comment]
I will close this as currently Source.single is slow.

**Pull request closed without merging**